File: Segundo_Curso/Segundo_Cuatri/PDOO/java/src/irrgarten/Player.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.ArrayList;

/**
 *
 * @author el_dramas
 */
public class Player {
    static private final String DEFAULT_NAME = "Player #";
    static private final int INVALID_POS = -1;
    
    static private int MAX_WEAPONS = 2;
    static private int MAX_SHIELD = 3;
    static private int INITIAL_HEALTH = 10;
    static private int HITS2LOSE = 3;
    
    private String name;
    private char number;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;
    private int consecutiveHits = 0;
    
    private ArrayList<Weapon> weapons = new ArrayList<Weapon>();
    private ArrayList<Shield> shields = new ArrayList<Shield>();
    
    public Player(char number, float intelligence, float strength){
        this.name = DEFAULT_NAME + number;
        this.number = number;
        this.intelligence = intelligence;
        this.strength = strength;
        this.col = INVALID_POS;
        this.row = INVALID_POS;
    }
    
    public void resurrect(){
        this.resetHits();
        this.health = INITIAL_HEALTH;
        ArrayList<Shield> newShields = new ArrayList<Shield>();
        ArrayList<Weapon> newWeapons = new ArrayList<Weapon>();
        this.shields = newShields;
        this.weapons = newWeapons;
    }
    
    public int getRow(){
        return row;
    }
    
    public int getCol(){
        return col;
    }
    
    public char getNumber(){
        return number;
    }
    
    public void setPos(int row, int col){
        this.row = row;
        this.col = col;
    }
    
    public boolean dead(){
        return health <= 0;
    }
    
    public Directions move(Directions directions, ArrayList<Directions> validMoves){
        throw new UnsupportedOperationException();        
    }
    
    public float attack(){
        return this.strength+this.sumWeapons();
    }
    
    public boolean defend(float receivedReward){
        return this.manageHit(receivedReward);
    }
    
    public String toString(){
        return "P[" + name + ", " + intelligence + ", " + strength + ", " + health + ",(" + row + ", " +col + ")]"; 
    }
    
    private void receiveWeapon(Weapon w){
        throw new UnsupportedOperationException();
    }
    
    private void receiveShield(Shield s){
        throw new UnsupportedOperationException();
    }
    
    private Weapon newWeapon(){
        float power = Dice.randomStrength();
        int durability = Dice.usesLeft();
        Weapon arma = new Weapon(power,durability);
        return arma;
    }
    
    private Shield newShield(){
        float protection = Dice.randomStrength();
        int durability = Dice.usesLeft();
        Shield shield = new Shield(protection,durability);
        return shield;
    }
    
    private float sumWeapons(){
        float fullAttack = 0;
        
        for (int i=0; i<weapons.size(); ++i){
            fullAttack += weapons.get(i).attack();
        }
        
        return fullAttack;
    }
    
    private float sumShield(){
        float fullProtection = 0;
        
        for (int i=0; i<shields.size(); ++i){
            fullProtection += shields.get(i).protect();
        }
        
        return fullProtection;
    }
    
    private float defensiveEnergy(){
        float defenseEnergy  = intelligence + this.sumShield();
        return defenseEnergy;
    }
    
    private boolean manageHit(float receivedAttack){
        throw new UnsupportedOperationException();        
    }
    
    private void resetHits(){
        this.consecutiveHits = 0;
    }
    
    private void gotWounded(){
        this.health--;
    }
    
    private void incConsecutiveHits(){
        this.consecutiveHits++;
    }
    
    
    
}

File: Segundo_Curso/Segundo_Cuatri/PDOO/ruby/Irrgarten/Player.rb
#encoding: UTF-8


require_relative 'Dice'
require_relative 'Weapon'
require_relative 'Shield'

module Irrgarten

    class Player
        @@DEFAULT_NAME = "Player #"
        @@INVALID_POS = -1 #TODO: Ver si lo puedes meter junto con el de monstruo en Labyrinth
        
        @@MAX_WEAPONS = 2
        @@MAX_SHIELDS = 3
        @@INITIAL_HEALTH = 10
        @@HITS2LOSE = 3

        def initialize(number,intelligence,strength)
            @name = @@DEFAULT_NAME
            @number = number
            @intelligence = intelligence
            @strength =strength
            @health = @INITIAL_HEALTH
            @row = @INVALID_POS
            @col = @INVALID_POS
            @consecutiveHits = 0 
            @weapons = []
            @shields = []
        end
        #getter row y col
        attr_reader :row
        attr_reader :col
        attr_reader :number
    
        def setPos(row,col)
            if row >= 0 && col >= 0 then
                @row = row
                @col = col
            end
        end


        def receiveWeapon(w)
            for i in 0..@weapons.size do 
                wi = @weapons[i]
                if(wi.discard)
                    @weapons.shift
                end
            end
            size = @weapons.size
            if size<@@MAX_WEAPONS then
                w = newWeapon
                @weapons.push(w)	
            end
        end

        def receiveShield(s)
            for i in 0..@shields.size do
                si = @shields[i]
                if(si.discard)
                    @shields.shift
                end
            end
            size = @shields.size
            if(size<@@MAX_SHIELDS)
                s = newShield
                @shields.push(s)
            end
        end

        def move(direction, validMoves)
            size = validMoves.size
            contained = find(direction,validMoves)
            if (size > 0) && !contained then
                firstElement = validMoves[0]
                return firstElement
            else 	
                return direction
            end 
        end

        def manageHit(receivedAttack)
            defense=defensiveEnergy
            if defense<receivedAttack then
                gotWounded
                incConsecutiveHits
            else
                resetHits
            end
            lose=true
            if (@consecutiveHits == @@HITS2LOSE)||(dead()) then
                resetHits
            else	
                lose=false
            end
            return lose
        end

        def resurrect
            weapons.clear()
            shields.clear()
            health = @INITIAL_HEALTH
            consecutiveHits = 0
        end

        def setPos(row,col)
            @row = row
            @col = col
        end

        def dead
            @health <= 0
        end

        def attack
            @strength + sumWeapons
        end

        def defend(receivedAttack)
            manageHit(receivedAttack)
        end

        def to_s
            str="#{@name}, #{@number}, #{@intelligence}, #{@strength}\n"
            str+= "Weapons: ["
            str += @weapons[0] unless @weapons.size==0

            for w in 1..@weapons.size do
                str += " - " + w.to_s
            end
            str += "]\n"
            str+= "Shields: ["
            str += @shields[0] unless @shields.size==0

            for sh in 1..@shields.size do
                str += " - " + sh.to_s
            end
            str += "]\n"
            str
                
        end
        def receiveReward
            wReward = Dice::Dice.weaponsReward
            sReward = Dice::Dice.shieldsReward
            for i in 0..wReward do
                wnew = newWeapon
                receiveWeapon(wnew)
            end
            for i in 0..sReward do
                snew = newShield
                receiveShield(snew)
            end
            extraHealth = Dice::Dice.healthReward
            @health += extraHealth
        end

        private  

        def newWeapon
            Weapon.new(Dice::Dice.randomStrength, Dice::Dice.usesLeft)
        end

        def newShield
            Shield.new(Dice::Dice.randomShield, Dice::Dice.usesLeft)
        end

        def defensiveEnergy
            @intelligence + sumShields
        end

        def resetHits
            @consecutiveHits = 0
        end

        def gotWounded
            @health -= 1
        end

        def incConsecutiveHits
            @consecutiveHits +=1
        end

        def sumWeapons
            sum=0
            for i in 0..@weapons.size do
                sum += @weapons[i].attack
            end
            sum
        end

        def sumShields
            sum=0
            for i in 0..@shields.size do
                sum += @shields[i].protect
            end
            sum
        end
    def find(element,array)
        found = false
        i = 0
        while !found && (i<array.size) 
            if array[i] == element then 
                found = true
            else 
                i += 1
            end
        end
        found	
    end
end


p = Player.new('45',0,0)

puts p.to_s

p.setPos(0,0)

end
